### PR TITLE
Reimplemented Filter => Cond rewrite to apply more conservatively

### PR DESCRIPTION
**Pull request by djspiewak**
We can now eliminate autojoins even in cases where common sources are being obscured by filters.  We do this by rewriting the `Filter` to a `Cond` within the autojoin heritage (we don't eliminate the node, obviously, so other descendants will continue to see it as a `Filter`).  This optimization applies recursively and can open the door to other collapsing.

I wish we had a `Scan` `MapFunc`; if we had that, I could do something similar with `Subset`.

### CODE REVIEW

**Comment by alissapajer** [comment]
I'll review this this afternoon.

**Comment by alissapajer** [comment]
@djspiewak there is another flaky couchbase test. I am all for skipping these when they flake. But I think we might ending skipping half the couchbase test suite because many different couchbase tests sometimes fail with "Failed loading test data". So at some point to maintain a verified couchbase connector we will just need to fix this. (See #2329.)

**Comment by djspiewak** [comment]
@alissapajer Yeah I agree.  Not a windmill we can tilt right now, but if it goes too much further in this direction, we'll probably just mass-disable couchbase until polyrepo.

File: connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala [hunk 1]
```diff
@@ -110,23 +110,26 @@ final class MinimizeAutoJoins[T[_[_]]: BirecursiveT: EqualT] private () extends
...
+        val fm2 = fm flatMap { i2 =>
+          if (i2 === i)
+            func.Cond(predicate.map(κ(i)), i.point[FreeMapA], func.Undefined[Int])
+          else
+            i2.point[FreeMapA]
+        }
 
-    /*fm flatMap {
-      case QSFilter(source, predicate) =>
-        // tune into 102.5 FM, The Source
```
Comment (alissapajer) [inline connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: oh noes we lost this radio joke :(

File: connector/src/test/scala/quasar/qscript/qsu/MinimizeAutoJoinsSpec.scala [hunk 2]
```diff
@@ -280,7 +280,23 @@ object MinimizeAutoJoinsSpec extends Qspec with TreeMatchers with QSUTTypes[Fix]
                 HoleF[Fix],
                 func.Undefined)))
       }
-    }.pendingUntilFixed
```
Comment (alissapajer) [inline connector/src/test/scala/quasar/qscript/qsu/MinimizeAutoJoinsSpec.scala]: 🎉

File: connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala (hunk 1 above)
Comment (alissapajer) [reply connector/src/main/scala/quasar/qscript/qsu/MinimizeAutoJoins.scala]: Every university has a radio station or two, right?

**Comment by alissapajer** [comment]
@djspiewak looks like we need to do #3276 in order to merge this (or probably any other PRs at this point).

**Comment by djspiewak** [comment]
Lets ignore couchbase failures (ie not hold up PRs) until it’s done then. I’ll PR it tomorrow.

**Pull request merged**